perspective_reddit_bot/config_test.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A class for checking and applying moderation rules."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import unittest
+
+import config
+
+
+
+class ConfigTest(unittest.TestCase):
+
+  def test_has_duplicates(self):
+    self.assertTrue(config._has_duplicates([1, 1]))
+    self.assertFalse(config._has_duplicates([1, 2]))
+
+  def test_duplicate_items(self):
+    self.assertEqual([1],
+                     config._duplicate_items([1, 1]))
+    self.assertEqual([1, 2],
+                     config._duplicate_items([1, 2, 1, 2]))
+    self.assertEqual([],
+                     config._duplicate_items([1, 2]))
+
+  def test_load_rules(self):
+    rule_records = [
+      { 'perspective_score': { 'm1': '> 0.5' },
+        'action': 'report' },
+      { 'perspective_score': { 'm1': '> 0.5', 'm2': '> 0.9' },
+        'action': 'noop' },
+      { 'perspective_score': { 'm3': '> 0.2' },
+        'action': 'noop',
+        'name': 'third_rule' },
+    ]
+    rules, models = config._load_rules(rule_records)
+    self.assertEqual(set(('m1', 'm2', 'm3')), models)
+    self.assertEqual(3, len(rules))
+    self.assertEqual('m1', rules[0].name)
+    self.assertEqual('m1__m2', rules[1].name)
+    self.assertEqual('third_rule', rules[2].name)
+
+  def test_load_rules_fails_on_empty(self):
+    with self.assertRaises(ValueError):
+      config._load_rules([])
+
+  def test_load_rules_fails_on_missing_required_fields(self):
+    rule_record_missing_action = { 'perspective_score': { 'm1': '> 0.1' } }
+    rule_record_missing_scores = { 'action': 'noop' }
```

Comment:
Isn't this also missing actions and vice versa? Shouldn't we make these two test cases a bit more specific?

Reply:
done